keps/sig-apps/4563-evacuation-api/README.md

```diff
@@ -0,0 +1,1194 @@
+which describes other issues and consequences that would be solved by the Evacuation API.
+
+The major issues are:
+
+1. Without extra manual effort, an application running with a single replica has to settle for
+   experiencing application downtime during the node drain. They cannot use PDBs with
+   `minAvailable: 1` or `maxUnavailable: 0`, or they will block node maintenance. Not every user
+   needs high availability either, due to a preference for a simpler deployment model, lack of
+   application support for HA, or to minimize compute costs. Also, any automated solution needs
+   to edit the PDB to account for the additional pod that needs to be spun to move the workload
+   from one node to another. This has been discussed in issue [kubernetes/kubernetes#66811](https://github.com/kubernetes/kubernetes/issues/66811)
+   and in issue [kubernetes/kubernetes#114877](https://github.com/kubernetes/kubernetes/issues/114877).
+2. Similar to the first point, it is difficult to use PDBs for applications that can have a variable
+   number of pods; for example applications with a configured horizontal pod autoscaler (HPA). These
+   applications cannot be disrupted during a low load when they have only pod. However, it is
+   possible to disrupt the pods during a high load without experiencing application downtime. If
+   the minimum number of pods is 1, PDBs cannot be used without blocking the node drain. This has
+   been discussed in issue [kubernetes/kubernetes#93476](https://github.com/kubernetes/kubernetes/issues/93476).
+3. Graceful deletion of DaemonSet pods is currently only supported as part of (Linux) graceful node
+   shutdown. The length of the shutdown is again not application specific and is set cluster-wide
+   (optionally by priority) by the cluster admin. This does not take into account
+   `.spec.terminationGracePeriodSeconds` of each pod and may cause premature termination of
+   the application. This has been discussed in issue [kubernetes/kubernetes#75482](https://github.com/kubernetes/kubernetes/issues/75482)
+   and in issue [kubernetes-sigs/cluster-api#6158](https://github.com/kubernetes-sigs/cluster-api/issues/6158).
+4. Descheduler does not allow postponing eviction for applications that are unable to be evicted
+   immediately. This can result in descheduling of incorrect set of pods. This is outlined in the
+   KEP [kubernetes-sigs/descheduler#1354](https://github.com/kubernetes-sigs/descheduler/pull/1354).
+
+### Goals
+
+- Introduce new `evacuation.coordination.k8s.io` API.
+- Introduce evacuation controller.
+
+### Non-Goals
+
+- Synchronizing of the evacuation status to the pod status.
+
+## Proposal
+
+We will introduce a new term called evacuation. This is a contract between the evacuation instigator,
+the evacuee, and the evacuator. The contract is enforced by the API and an evacuation controller.
+We can think of evacuation as a managed and safer alternative to eviction.
```

Comment:
Watch out for the risk of confusing end users.

We already have preemption and eviction and people confuse the two. Or three, because there are [two kinds of eviction](https://kubernetes.io/docs/concepts/scheduling-eviction/#pod-disruption). And there's disruption in the mix.

Do we want to rename [Scheduling, Preemption and Eviction](https://kubernetes.io/docs/concepts/scheduling-eviction/) to _Scheduling, Preemption, **Evacuation** and Eviction_?

Reply:
Good idea, I added a mention which of what kind of eviction I mean here.

> Do we want to rename [Scheduling, Preemption and Eviction](https://kubernetes.io/docs/concepts/scheduling-eviction/) to Scheduling, Preemption, Evacuation and Eviction?

Yes, I think we want to add a new concept there and generally update the docs once we have an alpha.

Reply:
I'm with Tim here.

Preemption vs Eviction is already quite confusing. And TBH, I couldn't fully understand what the "evacuation" is supposed to solve by reading the summary or motivation.

From Goals:
> Changes to the eviction API to support the evacuation process.

If this is already going to be part of the Eviction API, maybe it should be named as a form of eviction. Something like "cooperative eviction" or "eviction with ack" or something along those lines?

Reply:
I'm all for framing it as another type of eviction; we already have two, so the extra cognitive load for users is not so much a problem.

Reply:
We can bikeshed names for the API kind; I'd throw a few of my own into the hat:
- EvictionRequest
- PodEvictionRequest

Reply:
I have renamed the api to `EvictionRequest` to make the term recognizable. A minor disadvantage is that we have to clarify what type of eviction we mean if we say evict (API-initiate eviction, or EvictionRequest)

The rest of the renames are as follows:

```
Evacuation (noun) -> EvictionRequest / Eviction Process
evacuation (verb) -> request an eviction / terminate / evict / process eviction
Evacuator -> Interceptor
Evacuee -> Pod
Evacuator Class -> Interceptor Class
Evacuation Instigator -> Eviction Requester
Evacuation Controller -> Eviction Request Controller
ActiveEvacuatorClass -> ActiveInterceptorClass
ActiveEvacuatorCompleted -> ActiveInterceptorCompleted
EvacuationProgressTimestamp -> ProgressTimestamp
ExpectedEvacuationFinishTime -> ExpectedInterceptorFinishTime
EvacuationCancellationPolicy -> EvictionRequestCancellationPolicy
FailedEvictionCounter -> FailedAPIEvictionCounter
```